src/Http/Routing/src/Internal/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,442 @@
+                }
+            }
+            else if (typeof(IResult).IsAssignableFrom(method.ReturnType))
+            {
+                body = Expression.Call(methodCall, ResultWriteResponseAsync, HttpContextParameter);
+            }
+            else if (method.ReturnType == typeof(string))
+            {
+                body = Expression.Call(StringResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+            else
+            {
+                body = Expression.Call(JsonResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+
+            Func<object?, HttpContext, Task>? requestDelegate = null;
+
+            if (consumeBodyDirectly)
+            {
+                // We need to generate the code for reading from the body before calling into the delegate
+                var lambda = Expression.Lambda<Func<object?, HttpContext, object?, Task>>(body, TargetArg, HttpContextParameter, DeserializedBodyArg);
+                var invoker = lambda.Compile();
+                object? defaultBodyValue = null;
+
+                if (allowEmptyBody && bodyType!.IsValueType)
+                {
+                    defaultBodyValue = Activator.CreateInstance(bodyType);
+                }
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    object? bodyValue;
+
+                    if (allowEmptyBody && httpContext.Request.ContentLength == 0)
+                    {
+                        bodyValue = defaultBodyValue;
+                    }
+                    else
+                    {
+                        try
+                        {
+                            bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType!);
+                        }
+                        catch (IOException ex)
+                        {
+                            LogRequestBodyIOException(httpContext, ex);
+                            httpContext.Abort();
+                            return;
+                        }
+                    }
+
+                    await invoker(target, httpContext, bodyValue);
+                };
+            }
+            else if (consumeBodyAsForm)
+            {
+                var lambda = Expression.Lambda<Func<object?, HttpContext, Task>>(body, TargetArg, HttpContextParameter);
+                var invoker = lambda.Compile();
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    // Generating async code would just be insane so if the method needs the form populate it here
+                    // so the within the method it's cached
+                    try
+                    {
+                        await httpContext.Request.ReadFormAsync();
+                    }
+                    catch (IOException ex)
```

Comment:
Also catch InvalidDataException for bad forms. That one should result in a 400 rather than an abort. cc: @pranavkm

Reply:
Should we log InvalidDataExceptions at the debug level? I suspect you'd see those only when there's an actual bug in the client rather than a transient network issue. I agree it shouldn't result in an error log like an uncaught exception from app code though. How about info level?

Reply:
Yes it should log when it suppresses the exception. I think Debug is the right level if it's going to include the exception, which it should. You wouldn't want exceptions in the info logs.